a CLOUDSC version using Atlas GPU structures

sbrdar opened the pull request:
This is a version of Atlas GPU data structures for CLOUDSC which runs SCC, SCC_HOIST, SCC_K_CACHING kernels:
* the code is isolated in src/cloudsc_gpu_atlas and can be build with "--with-atlas ON --with-cuda ON" (on Nvidia) or "--with-atlas ON --with-hip ON" (on AMD GPUs) during the "./cloudsc-bundle build" command. The built executables carry 'atlas' in their names
* add support for packing of nD-array variables into an (n+1)D-array using atlas::MultiField
* allocate of hoist variables is left to atlas::pluto resource manager which implements various offload strategies: unpinned, pinned, pinned_pool, zero-copy, managed

sbrdar committed:
use eigen/3.4.0 to prevent redifinitions stemming from EIGEN_COMP_PGI when using CUDA

sbrdar committed:
use eigen/3.4.0 to prevent redifinitions stemming from EIGEN_COMP_PGI when using CUDA

sbrdar committed:
enable GPU for Atlas & temporary disable other GPU implementations

sbrdar committed:
enable GPU for Atlas & temporary disable other GPU implementations

sbrdar committed:
add Atlas GPU implementation

sbrdar committed:
add Atlas GPU implementation

sbrdar committed:
adapt to FIELD_API implementation

sbrdar committed:
adapt to FIELD_API implementation

sbrdar committed:
revert temporary disabled features

sbrdar committed:
revert temporary disabled features

sbrdar committed:
remove unnecessary code

sbrdar committed:
remove unnecessary code

sbrdar committed:
correct the field_api git version

sbrdar committed:
correct the field_api git version

sbrdar committed:
fix a bug to have identical results to FIELD_API dwarf version

sbrdar committed:
fix a bug to have identical results to FIELD_API dwarf version

sbrdar committed:
use memory pinning for NVHPC - same as the FIELD_API version

sbrdar committed:
use memory pinning for NVHPC - same as the FIELD_API version

sbrdar committed:
remove debugging print outs

sbrdar committed:
remove debugging print outs

sbrdar committed:
temporary disable atlas::Multifield until it is merged in the develop branch of Atlas

sbrdar committed:
temporary disable atlas::Multifield until it is merged in the develop branch of Atlas

sbrdar committed:
set ACC flags

sbrdar committed:
set ACC flags

sbrdar committed:
1) use atlas::FieldSet GPU-functions instead of relying on multiple calls to atlas::Field GPU-ones; 2) add scc-hoist-atlas

sbrdar committed:
1) use atlas::FieldSet GPU-functions instead of relying on multiple calls to atlas::Field GPU-ones; 2) add scc-hoist-atlas

sbrdar committed:
use a preprocessor switch for scc and scc-hoist

sbrdar committed:
use a preprocessor switch for scc and scc-hoist

sbrdar committed:
adapt cloudsc_gpu_atlas to use host memory pinning + device mapping in Atlas Field

sbrdar committed:
adapt cloudsc_gpu_atlas to use host memory pinning + device mapping in Atlas Field

sbrdar committed:
enable MultiField for Atlas data structures, switch temporary to an Atlas branch until -develop- becomes merged

sbrdar committed:
enable MultiField for Atlas data structures, switch temporary to an Atlas branch until -develop- becomes merged

sbrdar committed:
add MultiField implementation for GPU

sbrdar committed:
add MultiField implementation for GPU

sbrdar committed:
make scc-hoist-atlas use multifield

sbrdar committed:
make scc-hoist-atlas use multifield

sbrdar committed:
update README with the Atlas GPU features

sbrdar committed:
update README with the Atlas GPU features

sbrdar committed:
Merge branch 'nab_atlas_gpu_field' of github.com:ecmwf-ifs/dwarf-p-cloudsc into nab_atlas_gpu_field

sbrdar committed:
Merge branch 'develop' into nab_atlas_gpu_field

sbrdar committed:
switch to atlas plut branch; enable cuda in atlas with --with-cuda

sbrdar committed:
clean up and make compile with new atlas Field/FieldSet signature

wdeconinck committed:
new Lumi modules environment

wdeconinck committed:
Make run on atos without multifield for now

wdeconinck committed:
add gh arch (note private cmake installation)

wdeconinck committed:
Add NITER environment variable to check for memory pool benefits

sbrdar committed:
move kernel into a subroutine (needed for acc deviceptr on Lumi)

sbrdar committed:
remove unused included module

sbrdar committed:
add host-allocation, device-allocation timers in src/cloudsc_gpu_atlas

sbrdar committed:
use standard timer for atlas device allocation

sbrdar committed:
fix path to custom CMake on gh1 of ecmwf

sbrdar committed:
add k_caching and hoist kernels of cloudsc in cloudsc_gpu_atlas

sbrdar committed:
remove an unused file

sbrdar committed:
cosmetics in Atlas gpu version

sbrdar committed:
do not time the d2h offload, it is not compatible with -managed- and -zerocopy- which need to be touched on the host to account for d2h offload

sbrdar committed:
validation only done for the environment variable NITER=1

sbrdar committed:
use pluto device pool for hoisted variables

sbrdar committed:
fix cray compilation

sbrdar committed:
fix usage of envirnoment variable HOIST_POOL

sbrdar committed:
fix the bug of FIELD RANK NOT SUPPORTED for Atlas GPU validation

sbrdar committed:
use the new atlas::MultiField constructor

sbrdar committed:
build Atlas tests

sbrdar committed:
Merge branch 'develop' into nab_atlas_gpu_field

sbrdar committed:
make compile again on Atos and nvhpc 22.11

sbrdar committed:
Merge branch 'develop' into nab_atlas_gpu_field

sbrdar committed:
forgot to resolve the conflict in file

sbrdar committed:
fix conflict

sbrdar committed:
Merge branch 'develop' into nab_atlas_gpu_field

sbrdar committed:
fix Atlas gpu configuration

sbrdar committed:
switch to Atlas v0.42.0

sbrdar committed:
use new way of creating Atlas multifield

wdeconinck commented on src/cloudsc_gpu_atlas/dwarf_cloudsc_atlas.F90:
```diff
@@ -92,7 +92,7 @@ PROGRAM DWARF_CLOUDSC
 
   ! Validate the output against serialized reference data
-  IF (NITER == 1) CALL GLOBAL_ATLAS_STATE%VALIDATE(FSET, FSPACE, NGPTOTG)
+  ! IF (NITER == 1) CALL GLOBAL_ATLAS_STATE%VALIDATE(FSET, FSPACE, NGPTOTG)
```
This should not be disabled

sbrdar replied:
Enabled again to execute every time after the first iteration.

wdeconinck commented on arch/eurohpc/lumi/cray-gpu/17.0.1/env.sh:
```diff
@@ -1,4 +1,4 @@
-# (C) Copyright 1988- ECMWF.
+) Copyright 1988- ECMWF.
```
Revert this change

sbrdar replied:
Done. Thanks!

wdeconinck commented on arch/eurohpc/lumi/cray-gpu/17.0.1/env.sh:
```diff
@@ -8,12 +8,36 @@
+    fi
+  else
+    module load $1
+  fi
```
This change is not needed.

sbrdar replied:
Thanks! It has been removed.

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -0,0 +1,558 @@
+    INTEGER(KIND=JPIM), INTENT(IN) :: NLEV
+    INTEGER(KIND=JPIM), INTENT(IN) :: NPROMA
+    REAL(KIND=JPRB), INTENT(IN)   :: PTSPHY       ! Physics timestep
+    REAL(KIND=JPRB), CONTIGUOUS  :: PLCRIT_AER(:,:,:)
```
I suggest to add INTENT for this and following arguments

sbrdar replied:
Thanks. I have added the INTENTs.

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -0,0 +1,558 @@
+    NLEV = FSPACE%LEVELS()
+    NGPTOT = FSPACE%SIZE()
+
+    NGPBLKS = (NGPTOT / NPROMA) + MIN(MOD(NGPTOT,NPROMA), 1)
```
```suggestion
    NGPBLKS = FIELD%SHAPE(4)
```

sbrdar replied:
Good point to avoid inconsistencies. Thanks!

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -0,0 +1,558 @@
+  END SUBROUTINE CLOUDSC_KERNEL
+
+
+  SUBROUTINE CLOUDSC_DRIVER(FSET, NUMOMP, NGPTOTG, KFLDX, PTSPHY)
```
NGPTOTG, KFLDX are unused variables (except perhaps for a print) and can be refactored out.

Also not sure NUMOMP makes sense in the GPU context. Perhaps this should be forcefully set to 1, @reuterbal ?

sbrdar replied:
I have removed both unused arguments in the CLOUDSC_DRIVER subroutine.

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -0,0 +1,558 @@
+      READ(HOIST_POOL_STR(1:HOIST_POOL_STR_LEN), *) HOIST_POOL
+    ENDIF
+    IF (ALL(HOIST_POOL/=[-1,0,1])) THEN
+        PRINT *, "HOIST_POOL must be -1, 0, or 1. Exiting."
```
I'd remove the '-1' option for HOIST_POOL, and don't do the `acc enter data create`, and only use pluto.

sbrdar replied:
I have removed the '-1' option, we rely on 'pluto' to allocate the hoisting variables.

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -0,0 +1,558 @@
+    CHARACTER(32) :: HOIST_POOL_STR
+    INTEGER(KIND=JPIM) :: TID ! thread id from 0 .. NUMOMP - 1
+
+#ifdef CLOUDSC_GPU_SCC_HOIST
+    ! Local declarations of promoted temporaries via atlas::pluto
+    TYPE(pluto_allocator) :: device_allocator
+    REAL(KIND=JPRB), POINTER :: ZFOEALFA(:,:,:), ZTP1(:,:,:), &
+        & ZLI(:,:,:), ZA(:,:,:), ZAORIG(:,:,:), ZLIQFRAC(:,:,:), &
+        & ZICEFRAC(:,:,:), ZQSMIX(:,:,:), ZQSLIQ(:,:,:), &
+        & ZQSICE(:,:,:), ZFOEEWMT(:,:,:), ZFOEEW(:,:,:), ZFOEELIQT(:,:,:)
+    REAL(KIND=JPRB), POINTER :: ZQX(:,:,:,:), ZQX0(:,:,:,:), &
+        & ZPFPLSX(:,:,:,:), ZLNEG(:,:,:,:), ZQXN2D(:,:,:,:)
```
This can be factored out as a CLOUD_SCC_HOIST_DATA type, (including the TYPE(pluto_allocator))
Then the lines below to allocate and deallocate could be member functions of this type.

sbrdar replied:
This approach is nice but it drifts away from the rest of the code with i.e. FIELD_API. Perhaps we can think of it together with @reuterbal?

wdeconinck replied:
Sure you could leave it.

wdeconinck reviewed (changes requested):
Thanks @sbrdar for showing the power of atlas and pluto.
I have some suggestions.

sbrdar committed:
fix typos (PR #89)

sbrdar committed:
do not use acc to allocate hoist variables, but use pluto (PR #89)

sbrdar committed:
fixes according to PR #89

wdeconinck commented on src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:
```diff
@@ -266,6 +269,7 @@ SUBROUTINE CLOUDSC_GLOBAL_ATLAS_STATE_LOAD(SELF, FSET, FSPACE, NPROMA, NGPTOTG)
       MULTIFIELD_IN = atlas_MultiField(CONFIG)
       CALL FSET%ADD(MULTIFIELD_IN%FIELDSET()) 
       CALL MULTIFIELD_IN%FINAL()
+      PRINT *, "ERROR: MultiField not avaiable in Atlas/develop"
```
Correction, it is now :)

sbrdar replied:
Removed and also replaced the old usage of MultiField creation in the CPU version with the new one, similar to how it is done for the GPU version.

wdeconinck commented on src/cloudsc_gpu_atlas/dwarf_cloudsc_atlas.F90:
```diff
@@ -0,0 +1,118 @@
+USE ATLAS_MODULE
+USE, INTRINSIC :: ISO_C_BINDING
+USE ATLAS_FUNCTIONSPACE_BLOCKSTRUCTUREDCOLUMNS_MODULE
+USE ATLAS_MULTIFIELD_MODULE
```
```suggestion
```
can be removed.

sbrdar replied:
Thanks. It has been removed now.

sbrdar committed:
remove an unused performance timer

wdeconinck commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+        LIBS
+            atlas_f
+            pluto_f
+            cloudsc-common-${prec}-atlas
```
```suggestion
            cloudsc-common-atlas-${prec}
```
I'd assume that ${prec} should go at the end of the name. Of course also change everywhere used and defined.

sbrdar replied:
This is the design of @reuterbal to have libraries, like atlas etc, attached to cloudsc-common after the precision. If @reuterbal agrees to change it in his build system, I have no problem changing it in the above CMake file.

reuterbal replied:
Either is fine with me - I wasn't particularly strict about the naming scheme there, only trying to avoid any clashes and limiting the programming model exposure in the common libs.

sbrdar committed:
build Atlas GPU code only if CUDA or HIP are there

sbrdar committed:
also use new MultiField constructor for the CPU version with Atlas

sbrdar committed:
fix a seg error in validate function when using variable batching via atlas::MultiField

sbrdar committed:
cleanup

sbrdar committed:
1) remove unused arguments in the CLOUDSC_DRIVER subroutine; 2) add INTENT to the CLOUDSC_DRIVER kernel arguments

wdeconinck commented:
I would advise to reuse CPU and GPU codes where possible, to avoid code duplication and maintainability.
In particular, I think the validate, and creation of the fields and loading the data is probably identical.
The CPU OpenMP driver or GPU OpenACC driver can then be called independently.

sbrdar committed:
avoid duplicate code in cloudsc_fortran_atlas and cloudsc_gpu_atlas

sbrdar committed:
reuse the code in cloudsc_fortran_atlas/expand_atlas_mod

sbrdar commented:
> I would advise to reuse CPU and GPU codes where possible, to avoid code duplication and maintainability. In particular, I think the validate, and creation of the fields and loading the data is probably identical. The CPU OpenMP driver or GPU OpenACC driver can then be called independently.

I have now reused the expand_atlas_mod and validate_atlas_mod from 'cloudsc_fortran_atlas' in 'cloudsc_gpu_atlas'.

sbrdar committed:
clean up the old usage of MultiField

sbrdar committed:
fix compilation issues on lumi-g

reuterbal commented on arch/toolchains/ecmwf-hpc2020-nvhpc.cmake:
```diff
@@ -30,6 +30,8 @@ set( OpenMP_C_LIB_NAMES     "acchost" CACHE STRING "")
 # NB: We have to add `-mp` again to avoid undefined symbols during linking
 # (smells like an Nvidia bug)
 set( OpenACC_Fortran_FLAGS "-acc=gpu -mp=gpu -gpu=cc80,lineinfo,fastmath,gvmode" CACHE STRING "" )
+set( OpenACC_C_FLAGS "${OpenACC_Fortran_FLAGS}")
```
should this not be a CACHE variable, too?

sbrdar replied:
You are right, I did not find OpenACC_C_FLAGS in the cache list and it should be there. I am adding CACHE here.

reuterbal commented on bundle.yml:
```diff
@@ -104,6 +106,8 @@ options :
             DWARF-P-CLOUDSC_ENABLE_ACC={{value}}
             FIELD_API_ENABLE_ACC={{value}}
+            ATLAS_ENABLE_ACC={{value}}
+            ATLAS_ENABLE_CUDA={{value}}
```
Is CUDA required, or does it work also with only OpenACC? Since the CUDA support can be enabled separately via `--with-cuda`, it would make more sense to restrict this to ATLAS_ENABLE_ACC here if possible.

sbrdar replied:
Good point. CUDA is not required on --with-acc, Atlas can resolve now itself what it needs in the background.

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+ecbuild_add_option( FEATURE CLOUDSC_GPU_SCC_K_CACHING_ATLAS
+    DESCRIPTION "Build optimized GPU version of CLOUDSC using SCC with k-caching of local array in the kernel (based on Atlas + OpenACC)" DEFAULT ON
+    CONDITION atlas_FOUND AND (Serialbox_FOUND OR HDF5_FOUND)
+)
```
We have retired the per-variant ecbuild options because it did get a little out of hand, making it hard to spot what options or packages were selected or not found. Instead, we build all viable variants that are possible with the found dependencies. Since the include in `src/CMakeLists.txt` already ensures that Atlas is available, we can skip the ecbuild_add_option here and the `if (HAVE_CLOUDSC_GPU_SCC_ATLAS)`, `if(HAVE_CLOUDSC_GPU_SCC_HOIST_ATLAS)` etc. below.

sbrdar replied:
Indeed, cmake was more involved with condition testing. I am happy to clean up all add_option-s now, assuming the conditions are satisfied.

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+        SOURCES
+            cloudsc_global_atlas_state_mod.F90
+            ../cloudsc_fortran_atlas/expand_atlas_mod.F90
+            ../cloudsc_fortran_atlas/validate_atlas_mod.F90
```
I think we should maybe move these into the common lib at some point? 

In any case, since they are built for every target here, you get problems with aliasing in the .mod files and potential race conditions in parallel builds. Please try a build with Ninja, which is great at catching these issues:
```
ninja: build stopped: multiple rules generate cloudsc-dwarf/src/cloudsc_gpu_atlas.dp/cloudsc_global_atlas_state_mod.mod.
```

Therefore, as a minimum you will need to set per-target module directories here. See for example the relevant snippet from the GPU variants:

https://github.com/ecmwf-ifs/dwarf-p-cloudsc/blob/b42aacae409f3a7bec13bdf0f4c75bee5f5989ba/src/cloudsc_gpu/CMakeLists.txt#L35-L38

sbrdar replied:
A valuable comment, much appreciated. I have at the moment opted for the minimum after your suggestion, which seems to work.

reuterbal commented on src/CMakeLists.txt:
```diff
@@ -48,6 +48,9 @@ foreach( prec IN LISTS CLOUDSC_PRECISIONS )
   endif()
   if( HAVE_ATLAS )
     add_subdirectory( cloudsc_fortran_atlas cloudsc_fortran_atlas.${prec} )
+    if( HAVE_CUDA OR HAVE_HIP )
```
I'm wondering, are CUDA/HIP really required to build these variants? It's not performant but a CPU build should work and be runnable, no?

sbrdar replied:
It is true. We can compile and run Atlas GPU code for CPU. I have removed the CUDA/HIP requirement here, albeit without thoroughly testing.

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+      ${CMAKE_CURRENT_SOURCE_DIR}/../../config-files/input.h5 ${CMAKE_CURRENT_BINARY_DIR}/../../../input.h5 )
+  execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink
+      ${CMAKE_CURRENT_SOURCE_DIR}/../../config-files/reference.h5 ${CMAKE_CURRENT_BINARY_DIR}/../../../reference.h5 )
+endif()
```
This is redundant, please remove. Test data is being handled centrally in CMake now.

sbrdar replied:
Thanks. Cleaned up now.

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+#    target_link_options( dwarf-cloudsc-gpu-scc-atlas PUBLIC "-gpu=pinned" )
+#    target_link_options( dwarf-cloudsc-gpu-scc-hoist-atlas PUBLIC "-gpu=pinned" )
+#    target_link_options( dwarf-cloudsc-gpu-scc-k-caching-atlas PUBLIC "-gpu=pinned" )
+#endif()
```
Can be deleted?

sbrdar replied:
Thanks. Removed now.

reuterbal commented on arch/ecmwf/hpc2020/nvhpc/22.11/env.sh:
```diff
@@ -24,10 +24,12 @@ module_unload() {
 module purge
 
 # Load modules
+module_load cmake/3.25.2
```
CMake 3.28 is loaded further down
```suggestion
module_load cmake/3.25.2
```

sbrdar replied:
Indeed. I have removed 3.25.2 from the list.

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+        TARGET dwarf-cloudsc-gpu-scc-atlas
+        COMMAND bin/dwarf-cloudsc-gpu-scc-atlas
+        ARGS 1 1000 128
+        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/../../..
```
This is where the central input file mechanics will place the test inputs. Please change this also for the other test definitions in this file.
```suggestion
        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
```

sbrdar replied:
Done. Thanks.

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+    )
+
+    ecbuild_add_test(
+        TARGET dwarf-cloudsc-gpu-scc-hoist-atlas
+        COMMAND bin/dwarf-cloudsc-gpu-scc-hoist-atlas
```
```suggestion
        TARGET dwarf-cloudsc-gpu-scc-hoist-atlas-${prec}
        COMMAND bin/dwarf-cloudsc-gpu-scc-hoist-atlas-${prec}
```

sbrdar replied:
Cleaned up now. Thanks!

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+    )
+
+    ecbuild_add_test(
+        TARGET dwarf-cloudsc-gpu-scc-k-caching-atlas
+        COMMAND bin/dwarf-cloudsc-gpu-scc-k-caching-atlas
```
```suggestion
        TARGET dwarf-cloudsc-gpu-scc-k-caching-atlas-${prec}
        COMMAND bin/dwarf-cloudsc-gpu-scc-k-caching-atlas-${prec}
```

sbrdar replied:
... and this one as well.

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+    )
+
+    ecbuild_add_test(
+        TARGET dwarf-cloudsc-gpu-scc-atlas
+        COMMAND bin/dwarf-cloudsc-gpu-scc-atlas
```
Tests don't pass because target name doesn't match:
```suggestion
        TARGET dwarf-cloudsc-gpu-scc-atlas-${prec}
        COMMAND bin/dwarf-cloudsc-gpu-scc-atlas-${prec}
```

sbrdar replied:
... and this one as well.

reuterbal commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -0,0 +1,491 @@
+    INTEGER(KIND=JPIM) :: HOIST_POOL
+
+    HOIST_POOL = 0
+    CALL GET_ENVIRONMENT_VARIABLE("HOIST_POOL", VALUE=HOIST_POOL_STR, LENGTH=HOIST_POOL_STR_LEN)
```
Can you document this enviornment variable in the readme please?

sbrdar replied:
Thanks. Done. Also, added information on dwarf-cloudsc-gpu-scc-k-caching-atlas.

reuterbal commented on arch/ecmwf/gh/env.sh:
The env doesn't fully work on the Grace Hopper node, the prototype1 variant and the CUDA variants throw various compiler errors for me. Have you seen the same?

I'm ok with integrating it nonetheless since it been used to test the Atlas variant, but maybe add a print to the env.sh file that clearly states that the CUDA C and prototype1 variants are known to not work with this.

reuterbal reviewed (changes requested):
Many thanks for this, this is an exciting step forward and I'm very much looking forward to bringing this in! Apologies for the long review time, I did want to do this properly and really test this with the appropriate care.

First of all, the SCC and SCC-K-CACHING variants work flawlessly in single and double precision and also on larger problem sizes and produce very nice performance results! Well done!

- [ ] I left some remarks in-line, please address them. Most of the comments relate to the CMake integration, which needs some improvements. Also, I would like to make sure the tests work correctly and ideally switch Atlas on for them in a subsequent PR both in hpc-ci and at least on GH Actions configuration.

- [ ] For some reason, the per-thread throughput reporting (which should measure the kernel execution time) seems to not work correctly, e.g., on gh1 it simply shows 0:

```
$ bin/dwarf-cloudsc-gpu-scc-k-caching-atlas-dp 1 163840 128
Warning! Processor does not support trapping of floating-point exceptions.
 LMULTIFIELD =   F
 NGPTOTG     =        163840
 NGPTOT      =        163840
 NBLKS       =          1280
 NPROMA      =           128
 Host allocate [ms]:             9
 LMULTIFIELD =   F
 NGPTOTG     =        163840
 NGPTOT      =        163840
 NBLKS       =          1280
 NPROMA      =           128
 Host allocate [ms]:             1

### ITERATION 1/1
     NUMPROC=1, NUMOMP=1, NGTOT=163840, NGPTOTG=163840, NPROMA=128, NGPBLKS=1280
 Device allocate [ms]:            32
 Reference MFLOP count for 100 columns :  12.48232900
     NUMOMP    NGPTOT  #GP-cols     #BLKS    NPROMA tid# : Time(msec)  MFlops/s     col/s
          1    163840         0         0       128    0 :         21         0   7784519 @ rank#0:core#62
          1    163840         0      1280       128   -1 :         74    275665   2208443 : TOTAL @ rank#0
    1 x   1    163840         0      1280       128   -1 :         74    275665   2208443 : TOTAL
```

- [x] Finally, after fixing the issues mentioned in the comments, the following test doesn't work for me right now on AC GPUs (both in single and double precision and with and without OpenMP):

```
21/46 Test: dwarf-cloudsc-fortran-atlas-dp-serial
Command: "/.../nab_atlas_gpu_field/build/bin/dwarf-cloudsc-fortran-atlas-dp" "1" "100" "16"
Directory: /.../nab_atlas_gpu_field/build
"dwarf-cloudsc-fortran-atlas-dp-serial" start time: May 26 06:09 CEST
Output:
----------------------------------------------------------
  LMULTIFIELD:   T
(E) Exception: FieldSet: cannot find field "PLCRIT_AER"  (/.../nab_atlas_gpu_field/source/atlas/src/atlas/field/FieldSet.cc:120 field)

(E) =========================================
(E) TERMINATING dwarf-cloudsc-fortran-atlas-dp
(E) -----------------------------------------
(E) FieldSet: cannot find field "PLCRIT_AER"
(E) -----------------------------------------
(E) LOCATION:  (/.../nab_atlas_gpu_field/source/atlas/src/atlas/field/FieldSet.cc:120 field)
(E) -----------------------------------------
(E) BACKTRACE
(E) ------------------
```

And the SCC-HOIST test fails with a segfault:
```
24/46 Test: dwarf-cloudsc-gpu-scc-hoist-atlas-dp
Command: "/.../nab_atlas_gpu_field/build/bin/dwarf-cloudsc-gpu-scc-hoist-atlas-dp" "1" "1000" "128"
Directory: /.../nab_atlas_gpu_field/build/cloudsc-dwarf/src/cloudsc_gpu_atlas.dp/../../..
"dwarf-cloudsc-gpu-scc-hoist-atlas-dp" start time: May 26 06:09 CEST
Output:
----------------------------------------------------------
 LMULTIFIELD =   F
 NGPTOTG     =          1000
 NGPTOT      =          1000
 NBLKS       =             8
 NPROMA      =           128
 LMULTIFIELD =   F
 NGPTOTG     =          1000
 NGPTOT      =          1000
 NBLKS       =             8
 NPROMA      =           128

### ITERATION 1/1
     NUMPROC=1, NUMOMP=1, NGTOT=1000, NGPTOTG=1000, NPROMA=128, NGPBLKS=8
<end of output>
Test time =   0.38 sec
----------------------------------------------------------
Test Failed.
```

sbrdar committed:
1) make OpenACC_C_FLAGS cmake-cached; 2) clean module loading

sbrdar committed:
Atlas does not need CUDA enabled anymore

sbrdar committed:
1) build modules of each executable in a separate folder; 2) remove unnecessary checks for Atlas, HDF5, Serialbox. (thanks to Balthasar Reuter)

sbrdar committed:
update README with the HOIST_POOL variable

sbrdar committed:
forgot to disable check for CUDA/HIP in the GPU version, as the CPU version should run as well

sbrdar committed:
fix the CPU code to run correctly

sbrdar committed:
fix the env variable CLOUDSC_ATLAS_MULTIFIELD to take only 0 or 1 for simplicity

sbrdar committed:
identation and a fix for CMake > 4.0

sbrdar committed:
clean up

sbrdar committed:
a better match of variable allocation between fortran-atlas and gpu-atlas code versions

sbrdar committed:
forgot the file

sbrdar committed:
use SELF%KLEV only for the intialisation of FSPACE

reuterbal commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -7,119 +7,91 @@
+    ARGS 1 1000 128
+    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
+    OMP 1
+)
```
Sorry for not spotting this earlier. Can we add a second test that enables the hoist pool?
```suggestion
ecbuild_add_test(
    TARGET dwarf-cloudsc-gpu-scc-hoist-atlas-${prec}
    COMMAND bin/dwarf-cloudsc-gpu-scc-hoist-atlas-${prec}
    ARGS 1 1000 128
    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
    OMP 1
)
ecbuild_add_test(
    TARGET dwarf-cloudsc-gpu-scc-hoist-pool-atlas-${prec}
    COMMAND bin/dwarf-cloudsc-gpu-scc-hoist-atlas-${prec}
    ARGS 1 1000 128
    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
    OMP 1
    ENVIRONMENT HOIST_POOL=1
)
```

sbrdar replied:
Thanks for the hint. It has been added.

reuterbal commented on src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:
```diff
@@ -209,23 +212,21 @@ SUBROUTINE CLOUDSC_GLOBAL_ATLAS_STATE_LOAD(SELF, FSET, FSPACE, NPROMA, NGPTOTG)
 
-    LMULTIFIELD = .TRUE.
-    CALL GET_ENVIRONMENT_VARIABLE("CLOUDSC_ATLAS_MULTIFIELD",CENV,CENV_LEN)
+    CALL GET_ENVIRONMENT_VARIABLE("CLOUDSC_ATLAS_MULTIFIELD", CENV, CENV_LEN)
```
Would you mind also adding a second test with `CLOUDSC_ATLAS_MULTIFIELD=1`?

sbrdar replied:
... and this as well. Thank you!

reuterbal reviewed (changes requested):
Many thanks for addressing all the comments @sbrdar! This looks very good now.

I just left two small requests inline to add test variants for the options introduced via environment variables to make sure these features are tested appropriately.

One thing I noticed, however, is that when not specifying the `--with-cuda=ON` option, the atlas-scc and atlas-scc-hoist variants seem to segfault. But tests work fine with the `--with-cuda=ON` option. Tested with nvidia/24.5 on AC:

```sh
srun -c 64 ./cloudsc-bundle build --arch arch/ecmwf/hpc2020/nvhpc/24.5/ --with-atlas=ON --build-dir build.nvidia [--with-cuda=ON]
cd build.nvidia
source env.sh
srun -q dg -t 0:30:00 -n 1 -c 32 --mem=120G --gpus=1 ctest
```

Can you please have another look what might be the issue here?

sbrdar committed:
get a bug fixes version of Atlas (solves running 'cloudsc_gpu_atlas' on CPU)

sbrdar committed:
add other tests including using memory pool for hoisted variables and multifield

sbrdar committed:
note the compilation error on GraceHopper of ECMWF: prototype1

sbrdar committed:
output warning during configure

sbrdar committed:
fix atlas parts in the bundle configurations

sbrdar committed:
when running Atlas GPU code on CPU without CUDA/HIP fix: 1) FE_INVALID optimisation in the release build; 2) dwarf files for Atlas must not use openacc without Atlas CUDA/HIP (with Willem Deconinck <willem.deconinck@ecmwf.int>)

sbrdar commented:
> Many thanks for addressing all the comments @sbrdar! This looks very good now.
> 
> I just left two small requests inline to add test variants for the options introduced via environment variables to make sure these features are tested appropriately.
> 
> One thing I noticed, however, is that when not specifying the `--with-cuda=ON` option, the atlas-scc and atlas-scc-hoist variants seem to segfault. But tests work fine with the `--with-cuda=ON` option. Tested with nvidia/24.5 on AC:
> 
> ```shell
> srun -c 64 ./cloudsc-bundle build --arch arch/ecmwf/hpc2020/nvhpc/24.5/ --with-atlas=ON --build-dir build.nvidia [--with-cuda=ON]
> cd build.nvidia
> source env.sh
> srun -q dg -t 0:30:00 -n 1 -c 32 --mem=120G --gpus=1 ctest
> ```
> 
> Can you please have another look what might be the issue here?

We have not considered this so far, but it has been fixed now. A partial FE_INVALID check on the CLOUDSC function had to be omitted in the release build when running on CPU. Also, dwarf files using Atlas data should not use OpenACC if Atlas does not have CUDA/HIP enabled. OpenACC accesses Atlas data structured through the 'device_ptr' pragma call.

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -182,6 +183,10 @@ SUBROUTINE CLOUDSC_KERNEL( TIMER, NGPTOT, NGPBLKS, NLEV, NPROMA, PTSPHY, &
   call device_allocator%allocate(ZFOEELIQT, [NPROMA, NLEV, NGPBLKS])
 #endif
 
+!#if (ATLAS_HAVE_GPU == 0)
+    CALL IEEE_SET_HALTING_MODE(IEEE_INVALID, .FALSE.)
+!#endif
```
```suggestion
    LOGICAL :: LL_HALT_INVALID
    CALL IEEE_GET_HALTING_MODE(IEEE_INVALID, LL_HALT_INVALID)
    CALL IEEE_SET_HALTING_MODE(IEEE_INVALID, .FALSE.)
```
Also please write a little comment stating that this temporarily disables floating-point-trapping of FE_INVALID caused by too aggressive optimisation within the CLOUDSC function.

sbrdar replied:
Many thanks. It has been added.

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -253,6 +258,10 @@ SUBROUTINE CLOUDSC_KERNEL( TIMER, NGPTOT, NGPBLKS, NLEV, NPROMA, PTSPHY, &
 !$acc end parallel loop
 !$acc end data
 
+!#if (ATLAS_HAVE_GPU == 0)
+    CALL IEEE_SET_HALTING_MODE(IEEE_INVALID, .TRUE.)
+!#endif
```
```suggestion
    IF (LL_HALT_INVALID) CALL IEEE_SET_HALTING_MODE(IEEE_INVALID,  .TRUE.)
```
Only set it back to `.TRUE.` if it was `.TRUE.` before (`LL_HALT_INVALID`)

sbrdar replied:
Fixed. Thanks!

wdeconinck commented on bundle.yml:
```diff
@@ -73,6 +73,8 @@ projects :
         cmake   : >
             ATLAS_ENABLE_TESTS=OFF
             ATLAS_ENABLE_CUDA=OFF
+            ATLAS_ENABLE_HIP=OFF
+            ATLAS_ENABLE_ACC=OFF
```
CUDA, HIP, ACC are OFF by default, so no need to add here.

sbrdar replied:
Indeed, they are off in Atlas. Hence, removed here. Thanks!

wdeconinck commented on src/common/CMakeLists.txt:
```diff
@@ -131,14 +131,23 @@ endif()
+    set( USE_ACC ${HAVE_ACC} )
+    if ( NOT ATLAS_HAVE_GPU )
+       set( USE_ACC 0 )
+    endif()
```
Write a comment that Atlas was not compiled with GPU support, so that OpenACC instructions should be ignored in the dwarf

sbrdar replied:
Good point. I have added a cmake warning on disabling OpenACC instruction when GPU support has been disabled in Atlas.

sbrdar committed:
fix config: CUDA, HIP, ACC are off by default in Atlas

sbrdar committed:
revert FE_INVALID after disabling it temporarily

sbrdar committed:
cmake to issue a warning on the ACC change when GPU support is not enabled in Atlas

sbrdar commented:
> Many thanks for this, this is an exciting step forward and I'm very much looking forward to bringing this in! Apologies for the long review time, I did want to do this properly and really test this with the appropriate care.
> 
> First of all, the SCC and SCC-K-CACHING variants work flawlessly in single and double precision and also on larger problem sizes and produce very nice performance results! Well done!
> 
> * [ ]  I left some remarks in-line, please address them. Most of the comments relate to the CMake integration, which needs some improvements. Also, I would like to make sure the tests work correctly and ideally switch Atlas on for them in a subsequent PR both in hpc-ci and at least on GH Actions configuration.
> * [ ]  For some reason, the per-thread throughput reporting (which should measure the kernel execution time) seems to not work correctly, e.g., on gh1 it simply shows 0:
> 
> ```
> $ bin/dwarf-cloudsc-gpu-scc-k-caching-atlas-dp 1 163840 128
> Warning! Processor does not support trapping of floating-point exceptions.
>  LMULTIFIELD =   F
>  NGPTOTG     =        163840
>  NGPTOT      =        163840
>  NBLKS       =          1280
>  NPROMA      =           128
>  Host allocate [ms]:             9
>  LMULTIFIELD =   F
>  NGPTOTG     =        163840
>  NGPTOT      =        163840
>  NBLKS       =          1280
>  NPROMA      =           128
>  Host allocate [ms]:             1
> 
> ### ITERATION 1/1
>      NUMPROC=1, NUMOMP=1, NGTOT=163840, NGPTOTG=163840, NPROMA=128, NGPBLKS=1280
>  Device allocate [ms]:            32
>  Reference MFLOP count for 100 columns :  12.48232900
>      NUMOMP    NGPTOT  #GP-cols     #BLKS    NPROMA tid# : Time(msec)  MFlops/s     col/s
>           1    163840         0         0       128    0 :         21         0   7784519 @ rank#0:core#62
>           1    163840         0      1280       128   -1 :         74    275665   2208443 : TOTAL @ rank#0
>     1 x   1    163840         0      1280       128   -1 :         74    275665   2208443 : TOTAL
> ```
> 
>  * [x]  Finally, after fixing the issues mentioned in the comments, the following test doesn't work for me right now on AC GPUs (both in single and double precision and with and without OpenMP):
> 
> ```
> 21/46 Test: dwarf-cloudsc-fortran-atlas-dp-serial
> Command: "/.../nab_atlas_gpu_field/build/bin/dwarf-cloudsc-fortran-atlas-dp" "1" "100" "16"
> Directory: /.../nab_atlas_gpu_field/build
> "dwarf-cloudsc-fortran-atlas-dp-serial" start time: May 26 06:09 CEST
> Output:
> ----------------------------------------------------------
>   LMULTIFIELD:   T
> (E) Exception: FieldSet: cannot find field "PLCRIT_AER"  (/.../nab_atlas_gpu_field/source/atlas/src/atlas/field/FieldSet.cc:120 field)
> 
> (E) =========================================
> (E) TERMINATING dwarf-cloudsc-fortran-atlas-dp
> (E) -----------------------------------------
> (E) FieldSet: cannot find field "PLCRIT_AER"
> (E) -----------------------------------------
> (E) LOCATION:  (/.../nab_atlas_gpu_field/source/atlas/src/atlas/field/FieldSet.cc:120 field)
> (E) -----------------------------------------
> (E) BACKTRACE
> (E) ------------------
> ```
> 
> And the SCC-HOIST test fails with a segfault:
> 
> ```
> 24/46 Test: dwarf-cloudsc-gpu-scc-hoist-atlas-dp
> Command: "/.../nab_atlas_gpu_field/build/bin/dwarf-cloudsc-gpu-scc-hoist-atlas-dp" "1" "1000" "128"
> Directory: /.../nab_atlas_gpu_field/build/cloudsc-dwarf/src/cloudsc_gpu_atlas.dp/../../..
> "dwarf-cloudsc-gpu-scc-hoist-atlas-dp" start time: May 26 06:09 CEST
> Output:
> ----------------------------------------------------------
>  LMULTIFIELD =   F
>  NGPTOTG     =          1000
>  NGPTOT      =          1000
>  NBLKS       =             8
>  NPROMA      =           128
>  LMULTIFIELD =   F
>  NGPTOTG     =          1000
>  NGPTOT      =          1000
>  NBLKS       =             8
>  NPROMA      =           128
> 
> ### ITERATION 1/1
>      NUMPROC=1, NUMOMP=1, NGTOT=1000, NGPTOTG=1000, NPROMA=128, NGPBLKS=8
> <end of output>
> Test time =   0.38 sec
> ----------------------------------------------------------
> Test Failed.
> ```

Thank you very much @reuterbal for having a thorough look on Atlas part of the dwarf! Much appreciated!

In addition to my previous answers, I can add here:
1) the per-thread throughput reporting is back again;
2) the CPU code version in the folder 'cloudsc-fortran-atlas' has been fixed;
3) the GPU code version in the folder 'clodusc-gpu-atlas' has been fixed to also run on CPU;
4) if cmake omits the folder 'prototype1' and one or two more, you can compile the Atlas part of the dwarf on GraceHopper as well. I have checked that the code executes on GH200 correctly.

So far, re Atlas seems to work in intended configurations, but I might have missed a configuration of two which we do not support through our tests. I will be happy to look into a specific problems should such needed to be fixed.

Again, many thanks for your invaluable comments!

sbrdar commented:
> > I would advise to reuse CPU and GPU codes where possible, to avoid code duplication and maintainability. In particular, I think the validate, and creation of the fields and loading the data is probably identical. The CPU OpenMP driver or GPU OpenACC driver can then be called independently.
> 
> I have now reused the expand_atlas_mod and validate_atlas_mod from 'cloudsc_fortran_atlas' in 'cloudsc_gpu_atlas'.

Many thanks for review this PR, @wdeconinck ! Very insightful hints and remarks! Thank you!

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -183,9 +184,12 @@ SUBROUTINE CLOUDSC_KERNEL( TIMER, NGPTOT, NGPBLKS, NLEV, NPROMA, PTSPHY, &
-!#if (ATLAS_HAVE_GPU == 0)
+! temporary disable floating-point-trapping of FE_INVALID caused by
+! too aggressive optimisation within the CLOUDSC function
+#if (ATLAS_HAVE_GPU == 0)
```
I don't think the ATLAS_HAVE_GPU variable is defined at all. You should remove the #if, just keeping the IEEE_*_HALTING_MODE statements.

sbrdar replied:
Thanks. Removed now.

wdeconinck commented on src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90:
```diff
@@ -258,12 +262,12 @@ SUBROUTINE CLOUDSC_KERNEL( TIMER, NGPTOT, NGPBLKS, NLEV, NPROMA, PTSPHY, &
-
     CALL TIMER%THREAD_END(TID)
 
+#if (ATLAS_HAVE_GPU == 0)
```
Also this #if should go, but keep what's nested.

sbrdar replied:
.. and this one as well. Thanks.

sbrdar committed:
remove undefined preprocessor variable

sbrdar committed:
Merge branch 'develop' into nab_atlas_gpu_field

sbrdar committed:
add Atlas to the CI tests and build hpc tests for push on nab_atlas_gpu_field

sbrdar committed:
disable some of atlas tests

sbrdar committed:
treating unexpected value in CI yaml

sbrdar committed:
a separate CI matrix for Atlas

sbrdar committed:
fix build.yml

sbrdar committed:
remove extra include in yaml file

sbrdar committed:
add CI installation of Atlas, ecbuild, eckit, fckit

sbrdar committed:
fix CI

wdeconinck committed:
fix CI

wdeconinck commented:
linux-nvhpc ci fails with:
```
FAILED: [code=2] src/cloudsc_gpu_atlas.dp/CMakeFiles/dwarf-cloudsc-gpu-scc-atlas-dp.dir/cloudsc_driver_scc_mod.F90.o src/cloudsc_gpu_atlas.dp/module-dwarf-cloudsc-gpu-scc-atlas/cloudsc_driver_scc_mod.mod 
  /opt/nvhpc/Linux_x86_64/24.5/compilers/bin/nvfortran -I/home/runner/work/dwarf-p-cloudsc/dwarf-p-cloudsc/src/cloudsc_gpu_atlas -I/home/runner/work/dwarf-p-cloudsc/dwarf-p-cloudsc/src/common/include -I/home/runner/work/dwarf-p-cloudsc/dwarf-p-cloudsc/build/src/common.dp/module-atlas -I/home/runner/work/_temp/install/atlas/include -I/home/runner/work/_temp/install/atlas/module/atlas -I/home/runner/work/_temp/install/fckit/include -I/home/runner/work/_temp/install/fckit/module/fckit -I/home/runner/work/_temp/install/fckit/include/fckit -I/home/runner/work/_temp/deps/hdf5/include -Mlarge_arrays -g -O0 -Mbounds -traceback -module src/cloudsc_gpu_atlas.dp/module-dwarf-cloudsc-gpu-scc-atlas -fPIE -mp -c src/cloudsc_gpu_atlas.dp/CMakeFiles/dwarf-cloudsc-gpu-scc-atlas-dp.dir/cloudsc_driver_scc_mod.F90-pp.f90 -o src/cloudsc_gpu_atlas.dp/CMakeFiles/dwarf-cloudsc-gpu-scc-atlas-dp.dir/cloudsc_driver_scc_mod.F90.o
  NVFORTRAN-F-0004-Unable to open MODULE file pluto_module.mod (/home/runner/work/dwarf-p-cloudsc/dwarf-p-cloudsc/src/cloudsc_gpu_atlas/cloudsc_driver_scc_mod.F90: 30)
  NVFORTRAN/x86-64 Linux 24.5-1: compilation aborted
  ```
 Perhaps we need to link the target `dwarf-cloudsc-gpu-scc-atlas-${prec}` with `pluto_f` directly.

wdeconinck commented:
The runner OS for intel-classic should be an older ubuntu-22.04. intel-llvm with ubuntu-24.04 should be OK.

wdeconinck commented:
The ci for "linux gnu-14" is failing in atlas where OpenMP is seemingly detected by CMake with -fopenmp flag but then the C++ code cannot find the "omp.h" include file for some weird reason. To be investigated why...

wdeconinck committed:
The runner OS for intel-classic should be an older ubuntu-22.04

wdeconinck commented:
> Perhaps we need to link the target `dwarf-cloudsc-gpu-scc-atlas-${prec}` with `pluto_f` directly.

This was already the case, so the problem is in the export of pluto in install-interface.
This is fixed in atlas branch hotfix/0.43.1, to be released as 0.43.1 once these tests pass.

Please update in builds.yml "refs/tags/0.43.0" to "hotfix/0.43.1" to see. If all works, I will release 0.43.1 and then we can update the tag once more to "refs/tags/0.43.1"

wdeconinck committed:
Update build.yml

wdeconinck committed:
Update build.yml

wdeconinck committed:
Update build.yml

wdeconinck committed:
Update build.yml

try without quotes

wdeconinck committed:
Update build.yml

Remove ENABLE_ATLAS=ON as it is default ON

wdeconinck committed:
fix GET_THREAD_NUM in case where are not in openmp regime

wdeconinck commented:
It seems the OpenMP for the gnu-14 on ubuntu-24.04 is really problematic. Perhaps we need to turn off OMP feature for this entire build?

intel-classic build shows that in case of multifield the expand is creating temporaries. This is because the dimensions of the second argument field have contiguous dimensions whereas the field passed is discontiguous slice. Changing the dimensions from `(nproma, nlev, nblk)` to `(:,:,:)` will solve this.

```
 43/48 Test #43: dwarf-cloudsc-gpu-scc-k-caching-multifield-atlas-sp ....***Failed    1.02 sec
   LMULTIFIELD =  T
   NGPTOTG     =         1000
   NGPTOT      =         1000
   NBLKS       =            8
   NPROMA      =          128
   Host allocate [ms]:       -56598
  forrtl: warning (406): fort: (1): In call to EXPAND_R2, an array temporary was created for argument #2
  
  Image              PC                Routine            Line        Source             
  dwarf-cloudsc-gpu  0000000000443103  expand_atlas_mod_          87  expand_atlas_mod.F90
  dwarf-cloudsc-gpu  000000000043847F  cloudsc_global_at         176  cloudsc_global_atlas_state_mod.F90
  dwarf-cloudsc-gpu  00000000004DA241  MAIN__                     78  dwarf_cloudsc_atlas.F90
  dwarf-cloudsc-gpu  000000000042B6CD  Unknown               Unknown  Unknown
  libc.so.6          00007FC53F229D90  Unknown               Unknown  Unknown
  libc.so.6          00007FC53F229E40  __libc_start_main     Unknown  Unknown
  dwarf-cloudsc-gpu  000000000042B5E5  Unknown               Unknown  Unknown
  ```

sbrdar committed:
fix for multifield to avoid compiler copy from non-contiguous to an temporary contiguous memory

wdeconinck commented on src/common/module/expand_mod.F90:
```diff
@@ -234,11 +234,18 @@ subroutine expand_i1(buffer, field, nlon, nproma, ngptot, nblocks)
   end subroutine expand_i1
 
   subroutine expand_r1(buffer, field, nlon, nproma, ngptot, nblocks)
+    use omp_lib
```
@sbrdar no need to add "use omp_lib" unless you are calling actual routines instead of pragmas.

wdeconinck committed:
remove uneccessary omp_lib includes

sbrdar committed:
fixing omp issues

sbrdar committed:
uninitialised TID

sbrdar committed:
cleanup

wdeconinck committed:
Update build.yml

Use latest versions of software stack

wdeconinck committed:
Remove unused files

wdeconinck commented on src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:
```diff
@@ -243,78 +244,48 @@ SUBROUTINE CLOUDSC_GLOBAL_ATLAS_STATE_LOAD(SELF, FSET, FSPACE, NPROMA, NGPTOTG)
     ELSE
-      DO IVAR = 1, INVAR_SIZE
-          CALL FSET%ADD(FSPACE%CREATE_FIELD(NAME=TRIM(IN_VAR_NAMES(IVAR)),   KIND=ATLAS_REAL(JPRB)))
+      DO IVAR = 1, 24
```
I don't like the number 24 here. Can this be a more descriptive size variable?

sbrdar replied:
Good point. Thanks! We now use NUM_IN_MFIELDS and NUM_OUT_MFIELDS in cloudsc_fortran_atlas. Also, a similar change has been done in cloudsc_gpu_atlas.

wdeconinck commented on src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:
```diff
@@ -243,78 +244,48 @@ SUBROUTINE CLOUDSC_GLOBAL_ATLAS_STATE_LOAD(SELF, FSET, FSPACE, NPROMA, NGPTOTG)
     ELSE
-      DO IVAR = 1, OUTVAR_SIZE
-          CALL FSET%ADD(FSPACE%CREATE_FIELD(NAME=TRIM(OUT_VAR_NAMES(IVAR)),   KIND=ATLAS_REAL(JPRB), LEVELS=SELF%KLEV+1))
+      DO IVAR = 1, 14
```
I don't like the number 14 here. Can this be a more descriptive size variable?

sbrdar replied:
Done. See previous comment.

wdeconinck commented on src/cloudsc_gpu_atlas/CMakeLists.txt:
```diff
@@ -0,0 +1,143 @@
+# (C) Copyright 1988- ECMWF.
```
1988 -> 2025

sbrdar replied:
Done.

wdeconinck commented on arch/ecmwf/gh/toolchain.cmake:
```diff
@@ -0,0 +1,68 @@
+# (C) Copyright 1988- ECMWF.
```
1988 -> 2025

sbrdar replied:
Done.

wdeconinck commented on arch/ecmwf/gh/env.sh:
```diff
@@ -0,0 +1,62 @@
+#! /usr/bin/env bash
+# (C) Copyright 1988- ECMWF.
```
1988 -> 2025

sbrdar replied:
Done. Thanks!

wdeconinck committed:
Remove install-ecbuild

wdeconinck committed:
Update bundle.yml package versions

wdeconinck committed:
Update build-hpc.yml

sbrdar committed:
1) more descriptive variables; 2) code cleanup

sbrdar committed:
1) more descriptive variables; 2) code cleanup

sbrdar committed:
copyright year updated

sbrdar committed:
copyright year update

wdeconinck committed:
Update build-hpc.yml

exclude failing SP tests

wdeconinck committed:
Update build-hpc.yml

Speed up compilation for LUMI

wdeconinck committed:
Update build-hpc.yml

Speedup compilation for ac-gpu nvhpc

wdeconinck committed:
cleanup

wdeconinck commented:
All CI is green now! I guess now it's time to clean up (rebase etc) the branch a bit before merging it.
Perhaps @reuterbal can have another review round to double check and approve.

wdeconinck commented on .github/workflows/build-hpc.yml:
```diff
@@ -46,14 +47,14 @@ jobs:
               #SBATCH --nodes=1
-              #SBATCH --ntasks=1
-              #SBATCH --cpus-per-task=32
+              #SBATCH --ntasks-per-node=4
+              #SBATCH --cpus-per-task=64
```
This makes for speedy compilation, giving access to 256 threads on the hpc node.

wdeconinck commented on .github/workflows/build-hpc.yml:
```diff
@@ -76,7 +79,8 @@ jobs:
               #SBATCH --time=00:20:00
               #SBATCH --nodes=1
-              #SBATCH --ntasks-per-node=1
+              #SBATCH --ntasks-per-node=8
+              #SBATCH --cpus-per-task=7
```
This gives access to 56 threads to compile on the dev-g LUMI node.
Empirically I found 56 is the maximum one can get.

reuterbal replied:
Yes, that's correct - see the disclaimer here: https://docs.lumi-supercomputer.eu/runjobs/scheduled-jobs/lumig-job/

wdeconinck commented on .github/workflows/build-hpc.yml:
```diff
@@ -8,6 +8,7 @@ on:
     branches:
     - main
     - develop
+    - 'nab_atlas_gpu_field'
```
This should be removed before merging!

sbrdar replied:
Done.

wdeconinck commented on arch/ecmwf/gh/env.sh:
```diff
@@ -0,0 +1,62 @@
+module_load HDF5/1.14.3-NVHPC-24.1-CUDA-12.3.0
+module_load Boost/1.82.0-GCC-12.3.0
+
+export PATH=/perm/nawd/software/gh/cmake/3.30.3/bin:$PATH
```
This needs a more permanent solution!

sbrdar committed:
remove after all CI are successful on the branch

reuterbal reviewed (approved):
Many thanks, to me this looks very good! The TLC applied to the build-hpc workflow is also much appreciated
